### DISCUSSION TITLE
Python-Agent non-intrusive steup with Gunicorn+Flask can not report trace

### DISCUSSION THREAD

**Discussion by WinterWaterWarm** [apache/skywalking#9640] [Q&A]
component version
```python
apache-skywalking ==0.8.0
gunicorn==20.1.0
Flask==2.0.0
```

Flask App
```python
# main.py
from flask import Flask
app = Flask(__name__)

@app.route('/')
def hello_world():
    return 'Hello World!'

if __name__ == '__main__':
    app.run()
```

 Run your application
```bash
sw-python -d run gunicorn --access-logfile - -w 4 -b 127.0.0.1:8080 main:app
```

Appearance：
![apearance](https://user-images.githubusercontent.com/39685919/191182333-dfeed7e7-2171-463b-a048-79719a5a977b.png)

There is normal heartbeat but not trace-report while request come in
Did I miss some work?

**Comment by Superskyyy** [comment]
@WinterWaterWarm Hello, have you checked out this documentation page? https://skywalking.apache.org/docs/skywalking-python/next/en/setup/faq/how-to-use-with-uwsgi/ It's the same story for gunicorn. 

I admit this is a non-optimal solution (requires you to do some manual work) but in the meantime please check it out to fix the problem. Please let me know if the fix is effective otherwise we can look into it together.

**Reply by WinterWaterWarm** [reply]
Thanks for your suggestion!

Unlike uwsgi，gunicorn can report trace correctly in Legacy Setup
Because gunicorn‘s default mode is lazy-app：[look here](https://docs.gunicorn.org/en/stable/design.html?highlight=preload#how-many-threads)

Here is the verification：

Flask App
```python
# main.py
from flask import Flask
app = Flask(__name__)

from skywalking import agent, config
config.init(collector_address="127.0.0.1:11800", service_name="sayHello")
agent.start()

@app.route('/')
def hello_world():
    return 'Hello World!'

if __name__ == '__main__':
    app.run()
```

Run application

```python
gunicorn --access-logfile - -w 4 -b 127.0.0.1:8080 main:app
```
![result1](https://user-images.githubusercontent.com/39685919/191408700-90027802-ddf8-454b-99e1-d9d25c1a05f5.png)


Go back to the sw-python cli ，document page [https://skywalking.apache.org/docs/skywalking-python/next/en/setup/cli/](url) 
`sw-python run command` cannot work in gounicorn(default mode)，Is that a bug? or i miss some work, you think

**Reply by Superskyyy** [reply]
> Thanks for your suggestion!
> 
> Unlike uwsgi，gunicorn can report trace correctly in Legacy Setup Because gunicorn‘s default mode is lazy-app：[look here](https://docs.gunicorn.org/en/stable/design.html?highlight=preload#how-many-threads)
> 
> Here is the verification：
> 
> Flask App
> 
> ```python
> # main.py
> from flask import Flask
> app = Flask(__name__)
> 
> from skywalking import agent, config
> config.init(collector_address="127.0.0.1:11800", service_name="sayHello")
> agent.start()
> 
> @app.route('/')
> def hello_world():
>     return 'Hello World!'
> 
> if __name__ == '__main__':
>     app.run()
> ```
> 
> Run application
> 
> ```python
> gunicorn --access-logfile - -w 4 -b 127.0.0.1:8080 main:app
> ```
> 
> ![result1](https://user-images.githubusercontent.com/39685919/191408700-90027802-ddf8-454b-99e1-d9d25c1a05f5.png)
> 
> Go back to the sw-python cli ，document page [https://skywalking.apache.org/docs/skywalking-python/next/en/setup/cli/](url) `sw-python run command` cannot work in gounicorn(default mode)，Is that a bug? or i miss some work, you think

I think there's an underlying bug in how the CLI works and interacts with web servers (because it's not a standard Python start-up, but rather called from exec*). I tried to fix it a few months ago and didn't spot where it went wrong. If you have time, feel free to take a look at the CLI code for Python agent.

**Reply by WinterWaterWarm** [reply]
OK, switching to Legacy Setup can fix my problem,
I will also read the CLI code and hope i can provide help for this issue.
thanks for your answer!

**Reply by Superskyyy** [reply]
> OK, switching to Legacy Setup can fix my problem, I will also read the CLI code and hope i can provide help for this issue. thanks for your answer!

Thank you too!
